Test timeout on `TestStoreRangeMergeNonConsecutive`

```
E0904 22:35:39.316859 8476 storage/store.go:1581  storage/store.go:1580: got committed raft command for 3 but have no range with that ID: {RangeID:3 OriginNodeID:100000001 Cmd:{Get:<nil> Put:<nil> ConditionalPut:<nil> Increment:<nil> Delete:<nil> DeleteRange:<nil> Scan:<nil> EndTransaction:<nil> RangeLookup:<nil> HeartbeatTxn:<nil> GC:<nil> PushTxn:header:<timestamp:<wall_time:0 logical:37 > cmd_id:<wall_time:0 random:0 > key:"b" replica:<node_id:1 store_id:1 replica_id:1 > range_id:3 txn:<name:"storage/replica_command.go:1098 (*Replica).AdminSplit" key:"b" id:"M#\262\314\251\316G\364\261*v\220\262o4H" priority:692384610 isolation:SERIALIZABLE status:PENDING epoch:0 timestamp:<wall_time:0 logical:37 > orig_timestamp:<wall_time:0 logical:37 > max_timestamp:<wall_time:0 logical:37 > certain_nodes:<> Writing:true > read_consistency:CONSISTENT > pushee_txn:<name:"storage/replica_command.go:1098 (*Replica).AdminSplit" key:"b" id:"M#\262\314\251\316G\364\261*v\220\262o4H" priority:692384610 isolation:SERIALIZABLE status:PENDING epoch:0 timestamp:<wall_time:0 logical:37 > orig_timestamp:<wall_time:0 logical:37 > max_timestamp:<wall_time:0 logical:37 > certain_nodes:<> Writing:true > now:<wall_time:0 logical:68 > push_type:CLEANUP_TXN range_lookup:false  ResolveIntent:<nil> ResolveIntentRange:<nil> MergeResponse:<nil> TruncateLog:<nil> Lease:<nil> ReverseScan:<nil> Batch:<nil>}}
panic: test timed out after 30s

goroutine 1 [chan receive]:
testing.RunTests(0x12196c0, 0x174d0a0, 0xba, 0xba, 0x7dff01)
        /usr/local/go/src/testing/testing.go:561 +0x8ad
testing.(*M).Run(0xc820183ef8, 0x7e04b2)
        /usr/local/go/src/testing/testing.go:493 +0x70
github.com/cockroachdb/cockroach/util/leaktest.TestMainWithLeakCheck(0xc820183ef8)
        /home/thunder/go/path/src/github.com/cockroachdb/cockroach/util/leaktest/leaktest.go:34 +0x21
github.com/cockroachdb/cockroach/storage_test.TestMain(0xc820183ef8)
        /home/thunder/go/path/src/github.com/cockroachdb/cockroach/storage/main_test.go:35 +0x21
main.main()
        github.com/cockroachdb/cockroach/storage/_test/_testmain.go:432 +0x113

goroutine 17 [syscall, locked to thread]:
runtime.goexit()
        /usr/local/go/src/runtime/asm_amd64.s:1696 +0x1

goroutine 2818 [semacquire]:
sync.runtime_Syncsemacquire(0xc8207c20d0)
        /usr/local/go/src/runtime/sema.go:237 +0x201
sync.(*Cond).Wait(0xc8207c20c0)
        /usr/local/go/src/sync/cond.go:62 +0x9b
github.com/cockroachdb/cockroach/util/stop.(*Stopper).Quiesce(0xc8200d82a0)
        /home/thunder/go/path/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:231 +0x1da
github.com/cockroachdb/cockroach/util/stop.(*Stopper).Stop(0xc8200d82a0)
        /home/thunder/go/path/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:191 +0x25
github.com/cockroachdb/cockroach/storage_test.TestStoreRangeMergeNonConsecutive(0xc820086000)
        /home/thunder/go/path/src/github.com/cockroachdb/cockroach/storage/client_merge_test.go:271 +0x173e
testing.tRunner(0xc820086000, 0x174dee0)
        /usr/local/go/src/testing/testing.go:455 +0x98
created by testing.RunTests
        /usr/local/go/src/testing/testing.go:560 +0x86d
```
The test script (come from  @tschottdorf :) )
```
make test TESTFLAGS=-c PKG=./storage &&
while [ $? -eq 0 ]; do echo -n . && ./storage.test  -test.run TestStoreRangeMergeNonConsecutive -test.timeout 30s --verbosity=1 &> boom; done;
less boom
```

--- a/kv/dist_sender_server_test.go
+++ b/kv/dist_sender_server_test.go
@@ -311,7 +311,7 @@ func TestReverseScanWithSplitAndMerge(t *testing.T) {
 
 	// Case 2: encounter with range merge .
 	// Merge the range ["e", "g") and ["g", "\xff\xff") .
-	if err := db.AdminMerge("g"); err != nil {
+	if err := db.AdminMerge("e"); err != nil {
 		t.Fatal(err)
 	}
 	if rows, err := db.ReverseScan("d", "g", 0); err != nil {

--- a/storage/client_merge_test.go
+++ b/storage/client_merge_test.go
@@ -191,7 +191,8 @@ func TestStoreRangeMergeWithData(t *testing.T) {
 	}
 }
 
-// TestStoreRangeMergeLastRange verifies that merging the last range is a noop.
+// TestStoreRangeMergeLastRange verifies that merging the last range
+// fails.
 func TestStoreRangeMergeLastRange(t *testing.T) {
 	defer leaktest.AfterTest(t)
 	store, stopper := createTestStore(t)
@@ -199,18 +200,19 @@ func TestStoreRangeMergeLastRange(t *testing.T) {
 
 	// Merge last range.
 	args := adminMergeArgs(proto.KeyMin, 1, store.StoreID())
-	if _, err := store.ExecuteCmd(context.Background(), &args); err != nil {
-		t.Fatalf("merge of last range should be a noop: %s", err)
+	if _, err := store.ExecuteCmd(context.Background(), &args); !testutils.IsError(err, "cannot merge final range") {
+		t.Fatalf("expected 'cannot merge final range' error; got %s", err)
 	}
 }
 
-// TestStoreRangeMergeNonConsecutive attempts to merge two ranges
-// that are not on same store.
-func TestStoreRangeMergeNonConsecutive(t *testing.T) {
+// TestStoreRangeMergeNonCollocated attempts to merge two ranges
+// that are not on the same stores.
+func TestStoreRangeMergeNonCollocated(t *testing.T) {
 	defer leaktest.AfterTest(t)
-	t.Skip("TODO(bdarnell): fix this flaky test")
-	store, stopper := createTestStore(t)
-	defer stopper.Stop()
+	mtc := startMultiTestContext(t, 4)
+	defer mtc.Stop()
+
+	store := mtc.stores[0]
 
 	// Split into 3 ranges
 	argsSplit := adminSplitArgs(proto.KeyMin, []byte("d"), 1, store.StoreID())
@@ -236,37 +238,16 @@ func TestStoreRangeMergeNonConsecutive(t *testing.T) {
 		log.Errorf("split ranges keys are equal %q!=%q", rangeA.Desc().StartKey, rangeC.Desc().StartKey)
 	}
 
-	// Read all of the system keys touched by the split transactions to
-	// resolve intents.  If intents are left unresolved then the
-	// asynchronous resolution may happen after the call to RemoveRange
-	// below, reviving the range and breaking the test.
-	if _, err := store.DB().Scan(keys.LocalMax, keys.SystemMax, 1000); err != nil {
-		t.Fatal(err)
-	}
+	// Replicate the ranges to different sets of stores. Ranges A and C
+	// are collocated, but B is different.
+	mtc.replicateRange(rangeA.Desc().RangeID, 0, 1, 2)
+	mtc.replicateRange(rangeB.Desc().RangeID, 0, 1, 3)
+	mtc.replicateRange(rangeC.Desc().RangeID, 0, 1, 2)
 
-	// Remove range B from store and attempt to merge. This is a bit of a hack and leaves some
-	// internals in an inconsistent state, so we must re-add the range later.
-	// This is sufficient for now to generate the "ranges not collocated" error; if this changes
-	// in the future we could make this test more realistic by using a multiTestContext
-	// and ChangeReplicas to arrange two ranges onto different stores/nodes.
-	//
-	// Wait for the leader lease to ensure things are quiescent before removing the range.
-	// See #702 and TestStoreExecuteCmdOutOfRange.
-	// TODO(bdarnell): refactor this test to rebalance the range onto a separate node
-	// when this is supported.
-	rangeB.WaitForLeaderLease(t)
-	if err := store.RemoveReplica(rangeB); err != nil {
-		t.Fatal(err)
-	}
-
+	// Attempt to merge.
 	desc := rangeA.Desc()
 	argsMerge := adminMergeArgs(desc.StartKey, 1, store.StoreID())
 	if _, err := rangeA.AdminMerge(argsMerge, desc); !testutils.IsError(err, "ranges not collocated") {
 		t.Fatalf("did not got expected error; got %s", err)
-	}
-
-	// Re-add the range. This is necessary for a clean shutdown.
-	if err := store.AddReplicaTest(rangeB); err != nil {
-		t.Fatal(err)
 	}
 }

--- a/storage/replica_command.go
+++ b/storage/replica_command.go
@@ -1239,54 +1239,49 @@ func (r *Replica) splitTrigger(batch engine.Engine, split *proto.SplitTrigger) e
 	return nil
 }
 
-// AdminMerge extends the range to subsume the range that comes next in
-// the key space. The range being subsumed is provided in args.SubsumedRange.
-// The EndKey of the subsuming range must equal the start key of the
-// range being subsumed. The merge is performed inside of a distributed
+// AdminMerge extends this range to subsume the range that comes next in
+// the key space. The merge is performed inside of a distributed
 // transaction which writes the updated range descriptor for the subsuming range
 // and deletes the range descriptor for the subsumed one. It also updates the
 // range addressing metadata. The handover of responsibility for
 // the reassigned key range is carried out seamlessly through a merge trigger
 // carried out as part of the commit of that transaction.
-// A merge requires that the two ranges are collocate on the same set of replicas.
+// A merge requires that the two ranges are collocated on the same set of replicas.
 //
 // The supplied RangeDescriptor is used as a form of optimistic lock. See the
 // comment of "AdminSplit" for more information on this pattern.
-func (r *Replica) AdminMerge(args proto.AdminMergeRequest, desc *proto.RangeDescriptor) (proto.AdminMergeResponse, error) {
+func (r *Replica) AdminMerge(args proto.AdminMergeRequest, origLeftDesc *proto.RangeDescriptor) (proto.AdminMergeResponse, error) {
 	var reply proto.AdminMergeResponse
 
-	if desc.EndKey.Equal(proto.KeyMax) {
+	if origLeftDesc.EndKey.Equal(proto.KeyMax) {
 		// Merging the final range doesn't make sense.
 		return reply, util.Errorf("cannot merge final range")
 	}
 
-	// Lookup subsumed range. This really belongs inside the transaction
-	// for consistency, but it is important (for transaction record placement)
-	// that the first action inside the transaction is the conditional put
-	// to change the left descriptor's end key. We look up the descriptor
-	// here only to get the new end key and then repeat the lookup inside the
-	// transaction.
-	var newEndKey proto.Key
+	updatedLeftDesc := *origLeftDesc
+
+	// Lookup subsumed (right) range. This really belongs inside the
+	// transaction for consistency, but it is important (for transaction
+	// record placement) that the first action inside the transaction is
+	// the conditional put to change the left descriptor's end key. We
+	// look up the descriptor here only to get the new end key and then
+	// repeat the lookup inside the transaction.
 	{
-		subsumedRng := r.rm.LookupReplica(desc.EndKey, nil)
-		if subsumedRng == nil {
+		rightRng := r.rm.LookupReplica(origLeftDesc.EndKey, nil)
+		if rightRng == nil {
 			return reply, util.Errorf("ranges not collocated")
 		}
 
-		newEndKey = subsumedRng.Desc().EndKey
-		log.Infof("initiating a merge of %s into %s", subsumedRng, r)
+		updatedLeftDesc.EndKey = rightRng.Desc().EndKey
+		log.Infof("initiating a merge of %s into %s", rightRng, r)
 	}
-
-	// Init updated version of existing range descriptor.
-	updatedDesc := *desc
-	updatedDesc.EndKey = newEndKey
 
 	if err := r.rm.DB().Txn(func(txn *client.Txn) error {
 		// Update the range descriptor for the receiving range.
 		{
 			b := &client.Batch{}
-			desc1Key := keys.RangeDescriptorKey(updatedDesc.StartKey)
-			if err := updateRangeDescriptor(b, desc1Key, desc, &updatedDesc); err != nil {
+			leftDescKey := keys.RangeDescriptorKey(updatedLeftDesc.StartKey)
+			if err := updateRangeDescriptor(b, leftDescKey, origLeftDesc, &updatedLeftDesc); err != nil {
 				return err
 			}
 			// Commit this batch on its own to ensure that the transaction record
@@ -1297,33 +1292,36 @@ func (r *Replica) AdminMerge(args proto.AdminMergeRequest, desc *proto.RangeDesc
 		}
 
 		// Do a consistent read of the second range descriptor.
-		desc2Key := keys.RangeDescriptorKey(desc.EndKey)
-		var desc2 proto.RangeDescriptor
-		if err := txn.GetProto(desc2Key, &desc2); err != nil {
+		rightDescKey := keys.RangeDescriptorKey(origLeftDesc.EndKey)
+		var rightDesc proto.RangeDescriptor
+		if err := txn.GetProto(rightDescKey, &rightDesc); err != nil {
 			return err
 		}
 
 		// Verify that the two ranges are mergeable.
-		if !bytes.Equal(desc.EndKey, desc2.StartKey) {
-			return util.Errorf("ranges are not adjacent; %s != %s", desc.EndKey, desc2.StartKey)
+		if !bytes.Equal(origLeftDesc.EndKey, rightDesc.StartKey) {
+			// Should never happen, but just in case.
+			return util.Errorf("ranges are not adjacent; %s != %s", origLeftDesc.EndKey, rightDesc.StartKey)
 		}
-		if !bytes.Equal(desc2.EndKey, newEndKey) {
-			return util.Errorf("range changed during merge; %s != %s", desc2.EndKey, newEndKey)
+		if !bytes.Equal(rightDesc.EndKey, updatedLeftDesc.EndKey) {
+			// This merge raced with a split of the right-hand range.
+			// TODO(bdarnell): needs a test.
+			return util.Errorf("range changed during merge; %s != %s", rightDesc.EndKey, updatedLeftDesc.EndKey)
 		}
-		if !replicaSetsEqual(desc.GetReplicas(), desc2.GetReplicas()) {
+		if !replicaSetsEqual(origLeftDesc.GetReplicas(), rightDesc.GetReplicas()) {
 			return util.Errorf("ranges not collocated")
 		}
 
 		// Remove the range descriptor for the deleted range.
 		b := &client.Batch{}
-		b.Del(desc2Key)
+		b.Del(rightDescKey)
 
-		if err := mergeRangeAddressing(b, desc, &updatedDesc); err != nil {
+		if err := mergeRangeAddressing(b, origLeftDesc, &updatedLeftDesc); err != nil {
 			return err
 		}
 
 		// Update the RangeTree.
-		if err := DeleteRange(txn, b, desc2.StartKey); err != nil {
+		if err := DeleteRange(txn, b, rightDesc.StartKey); err != nil {
 			return err
 		}
 
@@ -1331,12 +1329,12 @@ func (r *Replica) AdminMerge(args proto.AdminMergeRequest, desc *proto.RangeDesc
 		// loop do it, in order to provide a merge trigger.
 		b.InternalAddCall(proto.Call{
 			Args: &proto.EndTransactionRequest{
-				RequestHeader: proto.RequestHeader{Key: updatedDesc.StartKey},
+				RequestHeader: proto.RequestHeader{Key: updatedLeftDesc.StartKey},
 				Commit:        true,
 				InternalCommitTrigger: &proto.InternalCommitTrigger{
 					MergeTrigger: &proto.MergeTrigger{
-						UpdatedDesc:     updatedDesc,
-						SubsumedRangeID: desc2.RangeID,
+						UpdatedDesc:     updatedLeftDesc,
+						SubsumedRangeID: rightDesc.RangeID,
 					},
 				},
 			},
@@ -1344,7 +1342,7 @@ func (r *Replica) AdminMerge(args proto.AdminMergeRequest, desc *proto.RangeDesc
 		})
 		return txn.Run(b)
 	}); err != nil {
-		return reply, util.Errorf("merge of range into %d failed: %s", desc.RangeID, err)
+		return reply, util.Errorf("merge of range into %d failed: %s", origLeftDesc.RangeID, err)
 	}
 
 	return reply, nil

--- a/storage/replica_command.go
+++ b/storage/replica_command.go
@@ -1255,45 +1255,67 @@ func (r *Replica) splitTrigger(batch engine.Engine, split *proto.SplitTrigger) e
 func (r *Replica) AdminMerge(args proto.AdminMergeRequest, desc *proto.RangeDescriptor) (proto.AdminMergeResponse, error) {
 	var reply proto.AdminMergeResponse
 
-	// Lookup subsumed range.
 	if desc.EndKey.Equal(proto.KeyMax) {
-		// Noop.
-		return reply, nil
-	}
-	subsumedRng := r.rm.LookupReplica(desc.EndKey, nil)
-	if subsumedRng == nil {
-		return reply, util.Errorf("ranges not collocated; migration of ranges in anticipation of merge not yet implemented")
-	}
-	subsumedDesc := subsumedRng.Desc()
-
-	// Make sure the range being subsumed follows this one.
-	if !bytes.Equal(desc.EndKey, subsumedDesc.StartKey) {
-		return reply, util.Errorf("Ranges that are not adjacent cannot be merged, %s != %s", desc.EndKey, subsumedDesc.StartKey)
+		// Merging the final range doesn't make sense.
+		return reply, util.Errorf("cannot merge final range")
 	}
 
-	// Ensure that both ranges are collocate by intersecting the store ids from
-	// their replicas.
-	if !replicaSetsEqual(subsumedDesc.GetReplicas(), desc.GetReplicas()) {
-		return reply, util.Errorf("The two ranges replicas are not collocate")
+	// Lookup subsumed range. This really belongs inside the transaction
+	// for consistency, but it is important (for transaction record placement)
+	// that the first action inside the transaction is the conditional put
+	// to change the left descriptor's end key. We look up the descriptor
+	// here only to get the new end key and then repeat the lookup inside the
+	// transaction.
+	var newEndKey proto.Key
+	{
+		subsumedRng := r.rm.LookupReplica(desc.EndKey, nil)
+		if subsumedRng == nil {
+			return reply, util.Errorf("ranges not collocated")
+		}
+
+		newEndKey = subsumedRng.Desc().EndKey
+		log.Infof("initiating a merge of %s into %s", subsumedRng, r)
 	}
 
 	// Init updated version of existing range descriptor.
 	updatedDesc := *desc
-	updatedDesc.EndKey = subsumedDesc.EndKey
-
-	log.Infof("initiating a merge of %s into %s", subsumedRng, r)
+	updatedDesc.EndKey = newEndKey
 
 	if err := r.rm.DB().Txn(func(txn *client.Txn) error {
 		// Update the range descriptor for the receiving range.
-		b := &client.Batch{}
-		desc1Key := keys.RangeDescriptorKey(updatedDesc.StartKey)
-		if err := updateRangeDescriptor(b, desc1Key, desc, &updatedDesc); err != nil {
+		{
+			b := &client.Batch{}
+			desc1Key := keys.RangeDescriptorKey(updatedDesc.StartKey)
+			if err := updateRangeDescriptor(b, desc1Key, desc, &updatedDesc); err != nil {
+				return err
+			}
+			// Commit this batch on its own to ensure that the transaction record
+			// is created in the right place (our triggers rely on this).
+			if err := txn.Run(b); err != nil {
+				return err
+			}
+		}
+
+		// Do a consistent read of the second range descriptor.
+		desc2Key := keys.RangeDescriptorKey(desc.EndKey)
+		var desc2 proto.RangeDescriptor
+		if err := txn.GetProto(desc2Key, &desc2); err != nil {
 			return err
 		}
 
+		// Verify that the two ranges are mergeable.
+		if !bytes.Equal(desc.EndKey, desc2.StartKey) {
+			return util.Errorf("ranges are not adjacent; %s != %s", desc.EndKey, desc2.StartKey)
+		}
+		if !bytes.Equal(desc2.EndKey, newEndKey) {
+			return util.Errorf("range changed during merge; %s != %s", desc2.EndKey, newEndKey)
+		}
+		if !replicaSetsEqual(desc.GetReplicas(), desc2.GetReplicas()) {
+			return util.Errorf("ranges not collocated")
+		}
+
 		// Remove the range descriptor for the deleted range.
-		// TODO(bdarnell): need a conditional delete?
-		desc2Key := keys.RangeDescriptorKey(subsumedDesc.StartKey)
+		b := &client.Batch{}
 		b.Del(desc2Key)
 
 		if err := mergeRangeAddressing(b, desc, &updatedDesc); err != nil {
@@ -1301,7 +1323,7 @@ func (r *Replica) AdminMerge(args proto.AdminMergeRequest, desc *proto.RangeDesc
 		}
 
 		// Update the RangeTree.
-		if err := DeleteRange(txn, b, subsumedDesc.StartKey); err != nil {
+		if err := DeleteRange(txn, b, desc2.StartKey); err != nil {
 			return err
 		}
 
@@ -1314,7 +1336,7 @@ func (r *Replica) AdminMerge(args proto.AdminMergeRequest, desc *proto.RangeDesc
 				InternalCommitTrigger: &proto.InternalCommitTrigger{
 					MergeTrigger: &proto.MergeTrigger{
 						UpdatedDesc:     updatedDesc,
-						SubsumedRangeID: subsumedDesc.RangeID,
+						SubsumedRangeID: desc2.RangeID,
 					},
 				},
 			},
@@ -1322,7 +1344,7 @@ func (r *Replica) AdminMerge(args proto.AdminMergeRequest, desc *proto.RangeDesc
 		})
 		return txn.Run(b)
 	}); err != nil {
-		return reply, util.Errorf("merge of range %d into %d failed: %s", subsumedDesc.RangeID, desc.RangeID, err)
+		return reply, util.Errorf("merge of range into %d failed: %s", desc.RangeID, err)
 	}
 
 	return reply, nil

